Feature Add TOTP

JASIM0021 opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

## Proposed changes
<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue below. -->
Add 2FA Management (TOTP ) to Mobile App

## Issue(s)	
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->
#6287



## Screenshots


https://github.com/user-attachments/assets/be4d2ef6-96cf-46c3-a7cf-05dc436e9de2



## Types of changes
<!-- What types of changes does your code introduce to Rocket.Chat? -->
<!-- Put an `x` in the boxes that apply -->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist
<!-- Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. -->

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [x] I have added necessary documentation (if applicable)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

JASIM0021 committed:
TOTA Setting [WIP] #6287

JASIM0021 committed:
TOTP [Done]  #6287

JASIM0021 committed:
test remove

JASIM0021 committed:
some changes

JASIM0021 committed:
email otp enable/disable added [Done]

JASIM0021 committed:
lint fix

diegolmello commented on app/containers/Model/CustomModel.tsx:
```diff
@@ -0,0 +1,49 @@
+
+	useEffect(() => {
+		if (open) {
+			Animated.timing(opacity, {
```
We should stop using React Native's Animated.
Always use react-native-reanimated.

diegolmello commented on app/containers/Model/CustomModel.tsx:
CustomModal, not Model

diegolmello commented on app/containers/Model/CustomModel.tsx:
```diff
@@ -0,0 +1,49 @@
+	};
+
+	return (
+		<Modal visible={open} transparent animationType='fade' onRequestClose={onClose} statusBarTranslucent>
```
TwoFactor uses react-native-modal https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/0d3b93f972f4313b22248807f60bbace3de729f0/app/containers/TwoFactor/index.tsx#L5

Why do you need this new component?

JASIM0021 replied:
I am developing a new CustomModal to provide more control and flexibility, allowing it to close when the user clicks any empty space outside of it. I am currently using import Modal from 'react-native-modal' in the same manner.

diegolmello commented on app/containers/Model/style.tsx:
```diff
@@ -0,0 +1,30 @@
+const styles = StyleSheet.create({
+	backdrop: {
+		flex: 1,
+		backgroundColor: 'rgba(0,0,0,0.5)',
```
Always use our theming system

diegolmello commented on app/views/SecurityPrivacyView/TotpModel/TotpEnableModel.tsx:
```diff
@@ -0,0 +1,137 @@
+import React, { useState } from 'react';
+import { View, Text, TouchableOpacity, TextInput } from 'react-native';
+import QRCode from 'react-native-qrcode-svg';
```
I don't think we should add QR code for this.
I saw there's the manual code there, but QR code is irrelevant.
We would require the user to scan with another device, which doesn't make sense.

JASIM0021 replied:
I will be removing the QR code section in the next pull request.

diegolmello commented on app/views/SecurityPrivacyView/TotpModel/TotpEnableModel.tsx:
```diff
@@ -0,0 +1,137 @@
+	};
+
+	return (
+		<CustomModal open={open} onClose={onClose}>
```
I'm pretty sure you can use react-navigation for this and our headers.

JASIM0021 replied:
I believe this component functions as a modal to display QR/MFA setup information. Given that, I don't think it's necessary to use react-navigation or include a header. Please feel free to correct me if I'm mistaken, @diegolmello.

JASIM0021 committed:
tota wip

JASIM0021 committed:
feat: Implement 2FA via email and TOTP

This commit introduces two-factor authentication (2FA) methods for enhanced security:

- **TOTP (Time-Based One-Time Password):**  Users can now enable TOTP using an authenticator app.
- **Email:** Users can enable 2FA with a code sent to their registered email.

Also,
- Display a success message on successful updates.
- Show a success message when backup codes are regenerated.
- Copy code Icon and copy manual code.

JASIM0021 committed:
Refactor: Improve Security Privacy View and Modal components

This commit includes the following changes:

- Import  with other reanimated imports in Modal component for better code readability and avoid ESLint errors.
- Remove unused variable  from state of SecurityPrivacyView.
- Call  without awaiting to avoid blocking UI thread.
- Replace  with  in TOTP methods to fix race condition.
- Remove unused  prop from  component.

JASIM0021 committed:
fix: Downgrade react-native-svg to v13.8.0

The react-native-svg package was downgraded to v13.8.0 and react-native-qrcode-svg has been removed due to conflicts and build issues in the newest version.  This resolves build errors on iOS.

JASIM0021 commented:
waiting for review

Rohit3523 commented:
We recently updated the React Native version to 0.77.2, and the PR has already been merged into the dev branch. Could you please pull the latest changes and test if everything is working fine? Thanks!

diegolmello commented on app/containers/Model/Modal.tsx:
```diff
@@ -0,0 +1,48 @@
+	children: React.ReactNode;
+}
+
+const Modal: React.FC<IModalProps> = ({ open, onClose, children }) => {
```
I'm not sure we need another Modal component.
Have you tested it on tablets?
Size might have to be changed to keep the same UI of the other components.

JASIM0021 replied:
You're right; it's not required.

diegolmello commented on app/containers/Model/style.tsx:
```diff
@@ -0,0 +1,34 @@
+import { Dimensions, Platform, StyleSheet } from 'react-native';
+
+const SCREEN_HEIGHT = Dimensions.get('window').height;
```
Not reactive.
It won't update if the user changes orientation.

diegolmello commented on app/containers/Model/style.tsx:
```diff
@@ -0,0 +1,34 @@
+		justifyContent: 'center'
+	},
+	modalContainer: {
+		backgroundColor: '#fff',
```
theme

JASIM0021 replied:
Done

diegolmello commented on app/containers/Model/style.tsx:
```diff
@@ -0,0 +1,34 @@
+		maxHeight: SCREEN_HEIGHT * 0.8,
+		...Platform.select({
+			ios: {
+				shadowColor: '#000',
```
theme

JASIM0021 replied:
Done

JASIM0021 committed:
Merge branch 'develop' of github.com:RocketChat/Rocket.Chat.ReactNative into feature/add-tfa-managment

JASIM0021 committed:
updated totp

JASIM0021 committed:
using custom icon

JASIM0021 commented:
| Dark Mode | Light Mode |
|-----------|------------|
| ![Dark Mode Screenshot](https://github.com/user-attachments/assets/e2d201a4-ca7e-4b9e-822a-00cd192a7107) | ![Light Mode Screenshot](https://github.com/user-attachments/assets/6d75d13d-bb32-4e3c-a3f2-88b7d73572e7) |

JASIM0021 committed:
repeated text rmove

JASIM0021 committed:
Merge branch 'develop' into feature/add-tfa-managment
